### PR TITLE
Port over odo parser

### PR DESCRIPTION
**Pull request by yangcao77**
### What does this PR do?
This PR moves the parser pkg from Odo repo to devfile/parser, and integrates with latest devfile go structs defined in devfile/api

### What issues does this PR fix or reference?
Fixes https://github.com/devfile/api/issues/183

### Is your PR tested? Consider putting some instruction how to test your changes
Tested by running the unit tests within this PR.

### CODE REVIEW

File: main.go
Comment (amisevsk) [inline main.go:50]:
```suggestion
```

File: main.go
Comment (amisevsk) [inline main.go:44]: Named returns aren't used here (`devfileoj`). We should either have
```go
func ParseDevfile(devfileLocation string) (devfile parser.DevfileObj, err error) {
//...
devfile, err = devfilepkg.ParseAndValidate(devfileLocation)
```
or
```go
func ParseDevfile(devfileLocation string) (parser.DevfileObj, error) {
//...
devfile, err := devfilepkg.ParseAndValidate(devfileLocation)
```

File: main.go
Comment (amisevsk) [inline main.go:46]: Not necessary to have this declaration on a separate line

File: pkg/devfile/parse.go
Comment (amisevsk) [inline pkg/devfile/parse.go:5]: Comments in import blocks break auto-formatting (e.g. by goimports)

File: pkg/devfile/parse.go
Comment (amisevsk) [inline pkg/devfile/parse.go:15]: The functions in this file are just wrappers on `parser.Parse[...]()` *without* validation. I'm not clear why they should exist (and why consumers wouldn't just do parsing and validation in two steps.

File: pkg/devfile/parser/configurables.go
Comment (amisevsk) [inline pkg/devfile/parser/configurables.go:189]: Consider using `fmt.Sprintf` here and a `[]String`, e.g.
```go
result = []String{"Available Parameters for Devfile Components:"}
for _, v := range GetDevfileSupportedParameters() {
  result = append(result, fmt.Sprintf("%s - %s", v, supportedDevfileParameterDescriptions[v]))
}
return strings.Join(result, "\n")
```

File: pkg/devfile/parser/data/common/command_helper.go
Comment (amisevsk) [inline pkg/devfile/parser/data/common/command_helper.go:20]: What's the difference between this and just inlining
```go
command.Id = strings.ToLower(command.Id)
```
where we would use this function?

File: pkg/devfile/parser/data/common/command_helper.go
Comment (amisevsk) [inline pkg/devfile/parser/data/common/command_helper.go:11]: Since `dc.Id` is exported, a getter is not needed (unless there's a plan to do some actual processing here in the future).

File: pkg/devfile/parser/data/common/component_helper.go
Comment (amisevsk) [inline pkg/devfile/parser/data/common/component_helper.go:9]: Could these functions be replaced with one?
```go
func getComponentType(component v1.Component) v1.ComponentType
```
and then usages could be
```go
type := getComponentType(myComponent)
if type == v1.ContainerComponentType {//...}
// or
if type == v1.VolumeComponentType {//...}
```
This would also allow you to use the result in a `switch` statement.

File: pkg/devfile/parser/data/common/project_helper.go
Comment (amisevsk) [inline pkg/devfile/parser/data/common/project_helper.go:11]: Named returns make sense here

File: pkg/devfile/parser/data/v2/header.go
Comment (amisevsk) [inline pkg/devfile/parser/data/v2/header.go:14]: While there are benefits to resolving the inlined struct in this way, it's probably not necessary -- we can just do
```
return d.Metadata
```
at which point this function seems unnecessary.

**Review by amisevsk — COMMENTED** [review]
In general I think it's a bad idea to have no-logic getters on exported fields (e.g. `return d.Parent`) unless we can justify more complicated logic in the future.

File: pkg/devfile/parse.go
Comment (yangcao77) [reply pkg/devfile/parse.go:14]: the validation will be uncommented once we have it ready in devfile/api

File: pkg/devfile/parser/data/common/component_helper.go
Comment (yangcao77) [reply pkg/devfile/parser/data/common/component_helper.go:11]: I have added `GetComponentType` function. 
IMO the `IsContainer` and `isVolume` or other component type can still be left as a helper function; it's makes consumers' life easier that they don't need to do extra check. 
WDYT?

File: pkg/devfile/parser/data/common/command_helper.go
Comment (yangcao77) [reply pkg/devfile/parser/data/common/command_helper.go]: I think you are right, except for they function has a return value which returning the new command.Id after setting it. 
this is how we use it in our code: https://github.com/devfile/parser/blob/3a759b5e0196f8cb5908054a83aa0537773874bf/pkg/testingutil/devfile.go#L84
But I'm fine with removing this function.

File: pkg/devfile/parser/configurables.go
Comment (yangcao77) [reply pkg/devfile/parser/configurables.go:189]: Good catch! This function is for odo, and should be removed from the generic parser repo. 
`GetDevfileSupportedParameters()` is enough for the library

**Comment by yangcao77** [comment]
@amisevsk they reason why we have those `Get()` functions is because information like `d.component` is not exposed in `devfileObj` which consumers will get after parsing.  They can only use interface defined in devfileData to access the struct. 
This is how the initial devfile parser was implemented, and is how odo currently using and implemented.

File: pkg/devfile/parser/data/common/component_helper.go
Comment (maysunfaisal) [inline pkg/devfile/parser/data/common/component_helper.go:48]: err should not be capital. I think we can also opt for a switch case here instead of if blocks

File: pkg/devfile/parse.go
Comment (amisevsk) [reply pkg/devfile/parse.go:14]: Makes sense, thanks.

File: pkg/devfile/parser/data/common/component_helper.go
Comment (amisevsk) [reply pkg/devfile/parser/data/common/component_helper.go:11]: Makes sense to me, especially for a library.

File: pkg/devfile/parser/context/fake.go
Comment (yangcao77) [reply pkg/devfile/parser/context/fake.go:1]: all fields in`DevfileCtx` are defined as internal, and cannot be accessible outside of the context package

File: pkg/devfile/parser/data/common/component_helper.go
Comment (yangcao77) [reply pkg/devfile/parser/data/common/component_helper.go]: These functions were port over from odo repo, and I'm fine with removing those log statements

File: pkg/devfile/parser/context/fake.go
Comment (wtam2018) [reply pkg/devfile/parser/context/fake.go:1]: You are right.  Let's move FakeContext() to xxx_test.go so that it is not visible outside of tests.   Also, we should consider not exporting this function.

File: pkg/devfile/parser/data/v2/common/component_helper.go
Comment (wtam2018) [inline pkg/devfile/parser/data/v2/common/component_helper.go]: Simplify to:
```
return component.Volume != nil
```

File: pkg/devfile/parser/data/v2/common/component_helper.go
Comment (wtam2018) [inline pkg/devfile/parser/data/v2/common/component_helper.go]: Simplify to:
```
return component.Container != nil
```

**Comment by kadel** [comment]
I don't thing that this is a good approach. 

I was under assumption that we have agreement on evolving and doing all necessary changes to parser in odo repo first, and then once we have independent parser package in odo we extract it from odo and put it here. 
If we start evolving parser package like this in a separate repo odo and this parser package will drift apart and we will never be able to switch to this. 

The first step was supposed to be in a the opposite direction. If there are additional things in devfile/parser that are not in odo it should be added into odo parser package.

**Comment by wtam2018** [comment]
When will odo parser finish with evolving?

**Comment by kadel** [comment]
> I don't thing that this is a good approach.
> 
> I was under assumption that we have agreement on evolving and doing all necessary changes to parser in odo repo first, and then once we have independent parser package in odo we extract it from odo and put it here.
> If we start evolving parser package like this in a separate repo odo and this parser package will drift apart and we will never be able to switch to this.
> 
> The first step was supposed to be in a the opposite direction. If there are additional things in devfile/parser that are not in odo it should be added into odo parser package.

I haven't seen this issue on odo side before https://github.com/openshift/odo/issues/4117#issue-722538678
It addresses all my concerns. I just wanted to make sure that we don't have the same logic in odo and in devfile/parser.
The [issue]( https://github.com/openshift/odo/issues/4117#issue-722538678) talks about using devfile/parser in odo.  We just have to make sure that it happens right after this PR is merged.

File: pkg/devfile/parser/context/fake.go
Comment (yangcao77) [reply pkg/devfile/parser/context/fake.go:1]: I agree we should rename the file to avoid confusion. 
The function should be kept as exported, because consumers(like odo) would need that function for their testing purpose.

**Comment by johnmcollier** [comment]
Should we be vendoring the go module dependencies here? Might be a good idea

File: pkg/devfile/parse.go
Comment (adisky) [inline pkg/devfile/parse.go:7]: why we need this here? I think it could be removed, later on we need to think how we want to expose library functions.

File: pkg/devfile/parser/data/v2/2.0.0/types.go
Comment (adisky) [inline pkg/devfile/parser/data/v2/2.0.0/types.go:4]: do we need this now? if not it can be removed

File: pkg/devfile/parser/types.go
Comment (adisky) [inline pkg/devfile/parser/types.go:11]: shouldn't this file be renamed to something else pkg/devfile/parser/types.go? and also this methods need to be implemented by devfileObj `DevfileObj` ?

File: pkg/devfile/parser/data/v2/2.0.0/types.go
Comment (yangcao77) [reply pkg/devfile/parser/data/v2/2.0.0/types.go:4]: We intentionally add this here for our future devfile 2.1 support. @maysunfaisal

File: pkg/devfile/parser/types.go
Comment (yangcao77) [reply pkg/devfile/parser/types.go:11]: I agree to rename the file. `types.go` is too confusing
This is how we implement in odo as well, I don't think we want to change the implementation.

Comment (adisky) [reply pkg/devfile/parser/types.go]: yes for now lets rename the file.

File: pkg/devfile/parser/data/v2/2.0.0/types.go
Comment (adisky) [reply pkg/devfile/parser/data/v2/2.0.0/types.go:4]: why we need our own structure in library, i think it will complicate things @maysunfaisal anything you have in mind that i am not able to understand.

File: main.go
Comment (yangcao77) [reply main.go:9]: I understood what do you mean. 
Actually this import alias was automatically added by vscode go plugin. I guess to make the imports neat and easy to view.

File: pkg/devfile/parser/data/v2/2.0.0/types.go
Comment (yangcao77) [reply pkg/devfile/parser/data/v2/2.0.0/types.go:4]: For minor version change of devfile schema version, they should be backward compatible, but might add additional fields in higher version. (in this case, let's say 2.1.0). They should share the same common methods defined in `/common` and `/v2`. but for 2.1.0, it might need specific functions, and might require different implementation on same function for 2.0 & 2.1. 
we define our own struct like that in order to support both 2.0 and 2.1. 
This is an example branch we used for testing, which 2.0 and 2.1 have different implementation on `GetCustomType()`: https://github.com/maysunfaisal/parser/commit/3fd444c23f259b55da551760efb4c5633585cda3

Comment (yangcao77) [reply pkg/devfile/parser/data/v2/2.0.0/types.go:4]: For now, the empty struct looks like meaning less; we just set it up for future 2.1 support.

Comment (maysunfaisal) [reply pkg/devfile/parser/data/v2/2.0.0/types.go:4]: @adisky Yes, the idea behind this is; there may be feature differences between 2.0.0 & 2.1.0. For example, 2.1.0 will be having build guidance which 2.0.0 is not going to support. Hence we laid out the dir 2.0.0 & 2.1.0 with their own structs and an  interface pkg/devfile/parser/data/v2/interface.go(empty for now) which will have the func definition GetBuildGuidance/AddBuildGuidance, etc where 2.0.0 and 2.1.0 will have their own implementation of the func.

This is exactly like the outermost `DevfileData` but one level nested to handle different implementation versions between 2.0.0 & 2.1.0. Check the New func in pkg/devfile/parser/data/v2/helper.go where we return either of the above structs to the new empty interface(which will eventually have funcs for new features in 2.1.0)

Comment (maysunfaisal) [reply pkg/devfile/parser/data/v2/2.0.0/types.go:4]: You may check this commit db34d9f4ecc1802426e7ae1962fbeb0f69e6f40f for a better idea of what the above paragraph just described

**Comment by amisevsk** [comment]
IMO if we're going to start vendoring dependencies in this repo, it should be done in a separate PR.

**Comment by yangcao77** [comment]
@amisevsk I'm fine with either way. I will modify the commit history if we want to remove the vendor folder. I don't want the git history to be too big. 
Is that the only concern from you?

@adisky And additional concerns?

I would like to get approval from you two before merging this PR. 
Let me know if you found difficulty reviewing with the PR due to the vendor folder. I will remove it if that's the case.

**Comment by adisky** [comment]
>  I don't want the git history to be too big.

@yangcao77 To keep the history clean you can squash all the commits before merging

> @adisky And additional concerns?
> 
> I would like to get approval from you two before merging this PR.
> Let me know if you found difficulty reviewing the PR due to the vendor folder. I will remove it if that's the case.

For Porting purpose from odo to here, It is good to go from my side but more work needed to make it consumable as library 
Logged two issues in devfile/api
https://github.com/devfile/api/issues/194
https://github.com/devfile/api/issues/193

**Comment by amisevsk** [comment]
If we're squashing this commit before merging then removing the vendoring is essential -- otherwise the commit that ends up in the history is completely useless from a history point of view. I have no problem with force-pushes to PR branches personally. 

Vendor vs not vendor is a topic worth discussing and not something we should slide into a PR about something generally separate.

**Comment by yangcao77** [comment]
I do not want to squash all commit into one single commit, that would lose all commit history inside this PR. 
Instead, I have removed the commits related to vendoring from commit history of the PR. 
Let me know if you have any other concerns. @amisevsk

**Comment by yangcao77** [comment]
I'm going to merge this PR if no additional concerns have been raised end of Tuesday(EDT). We need this PR to proceed our next step. Thanks.

**Review by maysunfaisal — APPROVED** [review]
All the review comments were addressed for the PR

/lgtm

**Pull request merged**
